pkg/cstor/migrate/pool.go

```diff
@@ -0,0 +1,324 @@
+	cvr "github.com/openebs/maya/pkg/cstor/volumereplica/v1alpha1"
+	deploy "github.com/openebs/maya/pkg/kubernetes/deployment/appsv1/v1alpha1"
+	pod "github.com/openebs/maya/pkg/kubernetes/pod/v1alpha1"
+	spc "github.com/openebs/maya/pkg/storagepoolclaim/v1alpha1"
+	"github.com/openebs/maya/pkg/util/retry"
+	"github.com/pkg/errors"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	types "k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	replicaPatch = `{
+	"spec": {
+		"replicas": 0
+	}
+}`
+	cspNameLabel           = "cstorpool.openebs.io/name"
+	cspUIDLabel            = "cstorpool.openebs.io/uid"
+	cspHostnameAnnotation  = "cstorpool.openebs.io/hostname"
+	cspiNameLabel          = "cstorpoolinstance.openebs.io/name"
+	cspiUIDLabel           = "cstorpoolinstance.openebs.io/uid"
+	cspiHostnameAnnotation = "cstorpoolinstance.openebs.io/hostname"
+	spcFinalizer           = "storagepoolclaim.openebs.io/finalizer"
+	cspcFinalizer          = "cstorpoolcluster.openebs.io/finalizer"
+)
+
+// Pool migrates the pool from SPC schema to CSPC schema
+func Pool(spcName, openebsNamespace string) error {
+
+	spcObj, migrated, err := getSPC(spcName, openebsNamespace)
+	if migrated {
+		klog.Infof("spc %s is already migrated to cspc", spcName)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	err = updateBDCLabels(spcName, openebsNamespace)
+	if err != nil {
+		return err
+	}
+	klog.Infof("Creating equivalent cspc for spc %s", spcName)
+	cspcObj, err := generateCSPC(spcObj, openebsNamespace)
+	if err != nil {
+		return err
+	}
+	err = updateBDCOwnerRef(cspcObj, openebsNamespace)
+	if err != nil {
+		return err
+	}
+	// List all cspi created with reconcile off
+	cspiList, err := cspi.NewKubeClient().
+		WithNamespace(openebsNamespace).
+		List(metav1.ListOptions{
+			LabelSelector: string(apis.CStorPoolClusterCPK) + "=" + cspcObj.Name,
+		})
+	if err != nil {
+		return err
+	}
+
+	// For each cspi perform the migration from csp that present was on
+	// node on which cspi came up.
+	for _, cspiObj := range cspiList.Items {
+		// Skip the migration if cspi is already in ONLINE state,
+		// which implies the migration is done and makes it idempotent
+		if cspiObj.Status.Phase != "ONLINE" {
+			err = csptocspi(&cspiObj, cspcObj, openebsNamespace)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	// Clean up old SPC resources after the migration is complete
+	err = spc.NewKubeClient().
+		Delete(spcName, &metav1.DeleteOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// getSPC gets the spc by name and verifies if the spc is already
+// migrated or not
+func getSPC(spcName, openebsNamespace string) (*apis.StoragePoolClaim, bool, error) {
+	spcObj, err := spc.NewKubeClient().
+		Get(spcName, metav1.GetOptions{})
+	// verify if the spc is already migrated.
+	if k8serrors.IsNotFound(err) {
+		klog.Infof("spc %s not found.", spcName)
+		_, err = cspc.NewKubeClient().
```

Comment:
If CSPC found then it doesn't mean that migration is successful there might be failure cases after creating CSPC. So IMO getting CSPI related to CSPC and verify the annotation `cstorpoolinstance.openebs.io/oldname` existence if there is no such annotation then return true(Or add annotation on CSPC openebs.io/migration: InProgress and remove this annotation or update value to Completed after successful completion of migration. Here it will be useful to  have single check).

Reply:
As discussed understood the point. Better to add in comments.